src/helpers/ring.rs

```diff
@@ -26,183 +27,270 @@ pub enum HelperAddr {
     Right,
 }
 
+/// Entry point to the system that enables MPC helpers to talk to each other. Every component
+/// that need message exchange to drive itself to completion calls methods on this trait with id
+/// argument that uniquely identifies its round of communication. Every helper executing the same
+/// protocol must use the same id.
+///
+/// `Ring` trait can be used to send and receive messages and that is what methods on this trait
+/// return. All messages will have the same protocol id
+pub trait CommunicationGateway {
+    type ChannelType: Ring;
+
+    /// Creates new communication channel for the given protocol id. All helpers are orchestrated
+    /// into a ring.
+    fn ring_channel(&self, id: ProtocolId) -> Self::ChannelType;
+}
+
 /// Trait for MPC helpers to communicate with each other. Helpers can send messages and
 /// receive messages from a specific helper.
 #[async_trait]
 pub trait Ring {
     /// Send message to the destination. Implementations are free to choose whether it is required
     /// to wait until `dest` acknowledges message or simply put it to a outgoing queue
-    async fn send<T: Message>(&self, dest: HelperAddr, msg: T) -> Result<(), Error>;
-    async fn receive<T: Message>(&self, source: HelperAddr) -> Result<T, Error>;
+    async fn send<T: Message>(&mut self, dest: HelperAddr, msg: T) -> Result<(), Error>;
+
+    /// Receive a given message from the `source`
+    async fn receive<T: Message>(&mut self, source: HelperAddr) -> Result<T, Error>;
 }
 
 #[cfg(test)]
 pub mod mock {
     use crate::helpers::error::Error;
-    use crate::helpers::ring::{HelperAddr, Message, Ring};
+    use crate::helpers::error::Error::{ReceiveError, SendError};
```

Comment:
use crate::helpers::error::Error::{self, ReceiveError, SendError};

Reply:
For my own education, what does adding "self" into a list of imports like this do?

Reply:
This:

```rust
use crate::helpers::error::Error::{self, ReceiveError, SendError};
```

...is equivalent to:

```rust
use crate::helpers::error::Error;
use crate::helpers::error::Error::{ReceiveError, SendError};
```

Reply:
Thanks!

Reply:
nice!